include/silo/common/range.h

```diff
@@ -0,0 +1,85 @@
+#pragma once
+
+#include <cassert>
+#include <cstddef>
+#include <string>
+
+#include <fmt/format.h>
+
+/// A range of values that support `-` yielding `size_t`, `+ (size_t)`
+/// yielding `T`, `<=`, `++` and `==` (todo: find and use the concepts
+/// for these).
+template <typename T>
+class Range {
+   T first_;
+   T beyond_last_;
+
+  public:
+   /// `first <= beyond_last` must be true.
+   Range(T first, T beyond_last)
+       : first_(first),
+         beyond_last_(beyond_last) {
+      assert(first <= beyond_last);
+   }
+
+   /// Get the first value of the Range. Unsafe, only valid if
+   /// !isEmpty().
+   T first() const { return first_; }
+
+   /// Get the value one after the last one in the Range.
+   T beyondLast() const { return beyond_last_; }
+
+   /// Mutate this Range to drop the first element. Unsafe, only
+   /// valid if !isEmpty().
+   void mutRest() { first_++; }
```

Comment:
Why is this method called `mutRest`? The name confuses me, but maybe I don't understand a concept?
From the code, I would have expected a name with something like "shift" or "increment".

Reply:
To distinguish from `rest` which returns a new `Range`; well there`s no `rest` method actually right now, but it's the same with `drop` and `mutDrop`. The "rest" naming is from functional languages, the rest of the list [ 12, 13, 14 ] is [ 13, 14 ]. I think "increment" would be ambiguous since it contains two positions, and I'm not sure what "shift" would mean.

The "mut" prefix could be "mutating" or "destructive" or "do" or anything else that indicates it's mutating the object in place.

Reply:
how about `incrementFirst`?

Reply:
> incrementing the first _element_ instead of the _index_ pointing to it

It's supposed to be a range, so a user should be aware how it works. At least to me it would be quite clear what "first" would refer to from reading the code. In the end we're not writing a library, but our own code where we have access to it for reading.
If that's unclear, we could also call the method something like "shift left boundary".

(I don't mind long method names as long as they make clear what it does)

Reply:
I've changed these things:

- changed the class to only have side-effect free methods, eliminating the `mut` methods.

- renamed `drop` to `skip` (functional languages generally use `drop` but that's what Rust calls it, although perhaps because Rust so predominantly uses `drop` for deleting objects)

- renamed `rest` to `skip1`, so the reader only has to keep in mind one base name, and I already renamed `inc` to `add1` so postfixing method names with 1 is a convention now.